### ISSUE TITLE
Unable to create native image with micronaut-http-client and jansi

### ISSUE THREAD

**Issue by auke-** [micronaut-projects/micronaut-core#5328]
I'm trying to build an interactive shell application with Micronaut and Jansi but i'm unable to compile it as a GraalVM native image. The compilation fails when both the `jansi` and `micronaut-http-client` dependency are on the classpath. The `jansi` dependency doesn't contain any links to the project which contains the class in the exception (netty-tcnative, io.netty.internal.tcnative.SSLPrivateKeyMethod) as far as I know but does use JNI which might trigger the error in Micronaut or Netty.

@ilopmar do you have any idea what could cause this error or how to make GraalVM show which class refers to to the mentioned class?

### Task List

- [x] Steps to reproduce provided
- [x] Stacktrace (if present) provided
- [ ] Example that reproduces the problem uploaded to Github
- [x] Full description of the issue provided (see below)

### Steps to Reproduce

1. Create a new Micronaut CLI app (`mn create-cli-app --build=maven --lang=java --test=junit --features=graalvm,jib,logback,mockito,shade my.package.app`)
2. Build native image (`./mvnw clean package -Dpackaging=docker-native -Pgraalvm`)
3. Add `jansi` dependency (`org.fusesource.jansi:jansi:2.3.2`, https://github.com/fusesource/jansi/)
4. Build native image (`./mvnw clean package -Dpackaging=docker-native -Pgraalvm`)
5. Change `micronaut-http-client` dependency from `test` to `compile` scope
6. Building native image will fail (`./mvnw clean package -Dpackaging=docker-native -Pgraalvm`)

### Expected Behaviour

The GraalVM native image compiler should succeed in creating a native image. This is the case when one of the `micronaut-http-client` or `jansi` dependencies are removed.

### Actual Behaviour

The GraalVM native image compiler fails with an exception:

```
Fatal error:java.lang.NoClassDefFoundError
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:490)
	at java.base/java.util.concurrent.ForkJoinTask.getThrowableException(ForkJoinTask.java:603)
	at java.base/java.util.concurrent.ForkJoinTask.get(ForkJoinTask.java:1006)
	at com.oracle.svm.hosted.NativeImageGenerator.run(NativeImageGenerator.java:499)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.buildImage(NativeImageGeneratorRunner.java:370)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.build(NativeImageGeneratorRunner.java:531)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner.main(NativeImageGeneratorRunner.java:119)
	at com.oracle.svm.hosted.NativeImageGeneratorRunner$JDK9Plus.main(NativeImageGeneratorRunner.java:568)
Caused by: java.lang.NoClassDefFoundError: io/netty/internal/tcnative/SSLPrivateKeyMethod
	at jdk.internal.vm.ci/jdk.vm.ci.hotspot.CompilerToVM.getDeclaredMethods(Native Method)
	at jdk.internal.vm.ci/jdk.vm.ci.hotspot.HotSpotResolvedObjectTypeImpl.getDeclaredMethods(HotSpotResolvedObjectTypeImpl.java:987)
	at com.oracle.svm.jni.access.JNIAccessibleMethod.anyMatchIgnoreReturnType(JNIAccessibleMethod.java:139)
	at com.oracle.svm.jni.access.JNIAccessibleMember.findHidingSubclasses(JNIAccessibleMember.java:103)
	at com.oracle.svm.jni.access.JNIAccessibleMember.findHidingSubclasses(JNIAccessibleMember.java:110)
	at com.oracle.svm.jni.access.JNIAccessibleMember.setHidingSubclasses(JNIAccessibleMember.java:83)
	at com.oracle.svm.jni.access.JNIAccessibleMethod.finishBeforeCompilation(JNIAccessibleMethod.java:135)
	at com.oracle.svm.jni.access.JNIAccessFeature.beforeCompilation(JNIAccessFeature.java:363)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$doRun$8(NativeImageGenerator.java:644)
	at com.oracle.svm.hosted.FeatureHandler.forEachFeature(FeatureHandler.java:71)
	at com.oracle.svm.hosted.NativeImageGenerator.doRun(NativeImageGenerator.java:644)
	at com.oracle.svm.hosted.NativeImageGenerator.lambda$run$2(NativeImageGenerator.java:495)
	at java.base/java.util.concurrent.ForkJoinTask$AdaptedRunnableAction.exec(ForkJoinTask.java:1407)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.ClassNotFoundException: io.netty.internal.tcnative.SSLPrivateKeyMethod
	at java.base/java.net.URLClassLoader.findClass(URLClassLoader.java:471)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:589)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
	... 18 more
```

### Environment Information

- **Operating System**: MacOS Catalina / Docker Desktop 3.3.1
- **Micronaut Version:** 2.4.2
- **JDK Version:** GraalVM Java 11 21.0.0.2 (same for 21.1.0)

### Example Application

Example application is simply created with steps to reproduce the issue.

**Comment by graemerocher** [comment]
Did you try add the `netty-tcnative` dependency?

**Comment by auke-** [comment]
Looks like `netty-tcnative` and `netty-tcnative-boringssl-static` aren't working out of the box when using the GraalVM native image compiler. The `netty-tcnative` dependency normally isn't necessary when using Micronaut http client in a GraalVM native image.

`io.netty:netty-tcnative:2.0.39.Final`:
```
Error: Unsupported features in 3 methods
Detailed message:
Error: Class initialization of io.netty.handler.ssl.OpenSslPrivateKeyMethod failed. Use the option --initialize-at-run-time=io.netty.handler.ssl.OpenSslPrivateKeyMethod to explicitly request delayed initialization of this class.
Original exception that caused the problem: java.lang.UnsatisfiedLinkError: 'int io.netty.internal.tcnative.NativeStaticallyReferencedJniMethods.sslSignRsaPkcsSha1()'
	at io.netty.internal.tcnative.NativeStaticallyReferencedJniMethods.sslSignRsaPkcsSha1(Native Method)
	at io.netty.internal.tcnative.SSLPrivateKeyMethod.<clinit>(SSLPrivateKeyMethod.java:22)
	at io.netty.handler.ssl.OpenSslPrivateKeyMethod.<clinit>(OpenSslPrivateKeyMethod.java:28)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1042)
	at jdk.unsupported/sun.misc.Unsafe.ensureClassInitialized(Unsafe.java:698)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.ensureClassInitialized(ConfigurableClassInitialization.java:174)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:607)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:127)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.maybeInitializeHosted(ConfigurableClassInitialization.java:165)
	at com.oracle.svm.hosted.SVMHost.initializeType(SVMHost.java:265)
	at com.oracle.graal.pointsto.meta.AnalysisType.lambda$new$0(AnalysisType.java:227)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Error: Class initialization of io.netty.internal.tcnative.CertificateVerifier failed. Use the option --initialize-at-run-time=io.netty.internal.tcnative.CertificateVerifier to explicitly request delayed initialization of this class.
Original exception that caused the problem: java.lang.UnsatisfiedLinkError: 'int io.netty.internal.tcnative.NativeStaticallyReferencedJniMethods.x509vOK()'
	at io.netty.internal.tcnative.NativeStaticallyReferencedJniMethods.x509vOK(Native Method)
	at io.netty.internal.tcnative.CertificateVerifier.<clinit>(CertificateVerifier.java:35)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1042)
	at jdk.unsupported/sun.misc.Unsafe.ensureClassInitialized(Unsafe.java:698)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.ensureClassInitialized(ConfigurableClassInitialization.java:174)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:607)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:584)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:584)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:127)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.shouldInitializeAtRuntime(ConfigurableClassInitialization.java:155)
	at com.oracle.svm.hosted.SVMHost.isInitialized(SVMHost.java:291)
	at com.oracle.graal.pointsto.meta.AnalysisType.isInitialized(AnalysisType.java:788)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.maybeEagerlyInitialize(BytecodeParser.java:4430)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genNewInstance(BytecodeParser.java:4628)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genNewInstance(BytecodeParser.java:4617)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genNewInstance(BytecodeParser.java:4612)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5409)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3436)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3243)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1109)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1003)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:76)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:212)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:223)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:357)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:313)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:302)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
	at com.oracle.graal.pointsto.flow.StaticInvokeTypeFlow.update(InvokeTypeFlow.java:434)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:547)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Error: Class initialization of io.netty.internal.tcnative.SSL failed. Use the option --initialize-at-run-time=io.netty.internal.tcnative.SSL to explicitly request delayed initialization of this class.
Original exception that caused the problem: java.lang.UnsatisfiedLinkError: 'int io.netty.internal.tcnative.NativeStaticallyReferencedJniMethods.sslOpCipherServerPreference()'
	at io.netty.internal.tcnative.NativeStaticallyReferencedJniMethods.sslOpCipherServerPreference(Native Method)
	at io.netty.internal.tcnative.SSL.<clinit>(SSL.java:67)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized0(Native Method)
	at java.base/jdk.internal.misc.Unsafe.ensureClassInitialized(Unsafe.java:1042)
	at jdk.unsupported/sun.misc.Unsafe.ensureClassInitialized(Unsafe.java:698)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.ensureClassInitialized(ConfigurableClassInitialization.java:174)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:607)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.computeInitKindAndMaybeInitializeClass(ConfigurableClassInitialization.java:127)
	at com.oracle.svm.hosted.classinitialization.ConfigurableClassInitialization.shouldInitializeAtRuntime(ConfigurableClassInitialization.java:155)
	at com.oracle.svm.hosted.SVMHost.isInitialized(SVMHost.java:291)
	at com.oracle.graal.pointsto.meta.AnalysisType.isInitialized(AnalysisType.java:788)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.maybeEagerlyInitialize(BytecodeParser.java:4430)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genInvokeStatic(BytecodeParser.java:1641)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.genInvokeStatic(BytecodeParser.java:1634)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBytecode(BytecodeParser.java:5406)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.iterateBytecodesForBlock(BytecodeParser.java:3436)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.processBlock(BytecodeParser.java:3243)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.build(BytecodeParser.java:1109)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.BytecodeParser.buildRootMethod(BytecodeParser.java:1003)
	at jdk.internal.vm.compiler/org.graalvm.compiler.java.GraphBuilderPhase$Instance.run(GraphBuilderPhase.java:84)
	at com.oracle.svm.hosted.phases.SharedGraphBuilderPhase.run(SharedGraphBuilderPhase.java:76)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:212)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:42)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.apply(Phase.java:38)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.parse(MethodTypeFlowBuilder.java:223)
	at com.oracle.graal.pointsto.flow.MethodTypeFlowBuilder.apply(MethodTypeFlowBuilder.java:357)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.doParse(MethodTypeFlow.java:313)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.ensureParsed(MethodTypeFlow.java:302)
	at com.oracle.graal.pointsto.flow.MethodTypeFlow.addContext(MethodTypeFlow.java:103)
	at com.oracle.graal.pointsto.flow.StaticInvokeTypeFlow.update(InvokeTypeFlow.java:434)
	at com.oracle.graal.pointsto.BigBang$2.run(BigBang.java:547)
	at com.oracle.graal.pointsto.util.CompletionExecutor.lambda$execute$0(CompletionExecutor.java:173)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1426)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
```

`io.netty:netty-tcnative-boringssl-static:2.0.38.Final`:
```
Error: Classes that should be initialized at run time got initialized during image building:
 io.netty.buffer.AbstractReferenceCountedByteBuf the class was requested to be initialized at run time (from the command line). To see why io.netty.buffer.AbstractReferenceCountedByteBuf got initialized use --trace-class-initialization=io.netty.buffer.AbstractReferenceCountedByteBuf
io.netty.buffer.UnpooledDirectByteBuf the class was requested to be initialized at run time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf). To see why io.netty.buffer.UnpooledDirectByteBuf got initialized use --trace-class-initialization=io.netty.buffer.UnpooledDirectByteBuf
io.netty.buffer.UnpooledHeapByteBuf the class was requested to be initialized at run time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf). To see why io.netty.buffer.UnpooledHeapByteBuf got initialized use --trace-class-initialization=io.netty.buffer.UnpooledHeapByteBuf
io.netty.buffer.ByteBufUtil the class was requested to be initialized at run time (from the command line). To see why io.netty.buffer.ByteBufUtil got initialized use --trace-class-initialization=io.netty.buffer.ByteBufUtil
io.netty.buffer.UnpooledUnsafeDirectByteBuf the class was requested to be initialized at run time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf). To see why io.netty.buffer.UnpooledUnsafeDirectByteBuf got initialized use --trace-class-initialization=io.netty.buffer.UnpooledUnsafeDirectByteBuf
io.netty.util.AbstractReferenceCounted the class was requested to be initialized at run time (from the command line). To see why io.netty.util.AbstractReferenceCounted got initialized use --trace-class-initialization=io.netty.util.AbstractReferenceCounted
io.netty.buffer.AbstractPooledDerivedByteBuf the class was requested to be initialized at run time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf). To see why io.netty.buffer.AbstractPooledDerivedByteBuf got initialized use --trace-class-initialization=io.netty.buffer.AbstractPooledDerivedByteBuf
io.netty.handler.ssl.ReferenceCountedOpenSslContext the class was requested to be initialized at run time (from the command line). To see why io.netty.handler.ssl.ReferenceCountedOpenSslContext got initialized use --trace-class-initialization=io.netty.handler.ssl.ReferenceCountedOpenSslContext
io.netty.handler.ssl.PemValue the class was requested to be initialized at run time (subtype of io.netty.util.AbstractReferenceCounted). To see why io.netty.handler.ssl.PemValue got initialized use --trace-class-initialization=io.netty.handler.ssl.PemValue
io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf the class was requested to be initialized at run time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf). To see why io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf got initialized use --trace-class-initialization=io.netty.buffer.UnpooledByteBufAllocator$InstrumentedUnpooledUnsafeDirectByteBuf
io.netty.buffer.PooledSlicedByteBuf the class was requested to be initialized at run time (subtype of io.netty.buffer.AbstractReferenceCountedByteBuf). To see why io.netty.buffer.PooledSlicedByteBuf got initialized use --trace-class-initialization=io.netty.buffer.PooledSlicedByteBuf
io.netty.buffer.ByteBufAllocator the class was requested to be initialized at run time (from the command line). To see why io.netty.buffer.ByteBufAllocator got initialized use --trace-class-initialization=io.netty.buffer.ByteBufAllocator
io.netty.handler.ssl.PemPrivateKey the class was requested to be initialized at run time (subtype of io.netty.util.AbstractReferenceCounted). To see why io.netty.handler.ssl.PemPrivateKey got initialized use --trace-class-initialization=io.netty.handler.ssl.PemPrivateKey
io.netty.buffer.PooledByteBufAllocator the class was requested to be initialized at run time (from the command line). To see why io.netty.buffer.PooledByteBufAllocator got initialized use --trace-class-initialization=io.netty.buffer.PooledByteBufAllocator
```

**Comment by ilopmar** [comment]
@auke- no, sorry, I don't know why they don't work together.

**Comment by auke-** [comment]
The issue is the same for Micronaut 2.4.4 with GraalVM 21.1.0.

I might have to ask this in the GraalVM channels, but do you know how to analyse such an issue?

**Comment by auke-** [comment]
Looks like you ran into the same issue with Tomcat: https://bz.apache.org/bugzilla/show_bug.cgi?id=64875

**Comment by ilopmar** [comment]
What happened with Tomcat is that they included the JNI GraalVM configuration by default and that broke everything. What they did to fix it was including the configuration file but not enabling it by default, so users can decide whether they want that JNI config or not.
For this case maybe the problem is that jansi is defining this: https://github.com/fusesource/jansi/blob/master/src/main/resources/META-INF/native-image/jansi/resource-config.json#L4

You may want to remove that line, but jansi locally and try again. Maybe this is something worth to report there too.

**Comment by auke-** [comment]
If I understand correctly the error isn't in the jni-config.json, but probably in the Netty code (or maybe Micronauts GraalVM configuration). The Micronaut application is working fine with `jansi`, even when compiled as a GraalVM native image. The error occurs when the `micronaut-http-client` dependency is added (which of course drags in some Netty dependencies).

When I remove the `jni-config.json` file the compilation succeeds even with the `micronaut-http-client` dependency, but the resulting application won't work since `jansi` needs those JNI calls.

I think the issues also can be reproduced without `jansi` by including any `jni-config.json` file, since `jansi` isn't related to Netty in any way.

**Comment by ilopmar** [comment]
We've been troubleshooting an issue this morning related with JNI and AWS, and we've reached the same conclusion. Somehow the flag `--allow-incomplete-classpath` doesn't work with GraalVM and JNI because the generation of the native image fails with `NoClassDefFoundError`.

I've opened this issue https://github.com/oracle/graal/issues/3387

**Comment by auke-** [comment]
I have tried to work around this issue by applying substitutions similar to the [substitutions used in Quarkus](https://github.com/quarkusio/quarkus/blob/main/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/graal/NettySubstitutions.java) but it keeps complaining about the missing classes `org.conscrypt.BufferAllocator` and `org.conscrypt.AllocatedBuffer`. That might be because I don't know how to replace or remove `io.netty.handler.ssl.ConscryptAlpnSslEngine` properly.

**Comment by ilopmar** [comment]
I've just tried those SSL substitutions in the sample app used in the GraalVM issue and it works!

I've also tried with a new Micronaut 2.4.4 application adding the jansi dependency and with the same substitutions, it works. There is an error with class `org/grails/datastore/mapping/validation/ValidationException` so I tried using a locally built Micronaut snapshot without that class and, as mentioned it works.

P.S: I've tried with a normal Micronaut app and used Gradle, but I think that won't make any difference.

**Comment by auke-** [comment]
Yes, you are right. I had to remove some stale `--initialize-at-run-time` lines from my previous attempts.

**Comment by traycho** [comment]
@auke Did you solve missing `org.conscrypt.BufferAllocator`  ?
@ilopmar  Do you plan a fix for this ? 
 I could confirm this is reproducible even with micronaut 2.1.0 and graalvm 20.3.0

**Comment by ilopmar** [comment]
@traycho you have a workaround here: https://github.com/micronaut-projects/micronaut-core/issues/5328#issuecomment-829912994 but we are waiting on GraalVM issue https://github.com/oracle/graal/issues/3387

**Comment by traycho** [comment]
> @traycho you have a workaround here: [#5328 (comment)](https://github.com/micronaut-projects/micronaut-core/issues/5328#issuecomment-829912994) but we are waiting on GraalVM issue [oracle/graal#3387](https://github.com/oracle/graal/issues/3387)

@ilopmar  Did you use all of the substitutions from this file  https://github.com/quarkusio/quarkus/blob/main/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/graal/NettySubstitutions.java?  Are you able to share your substitutions file ?

**Comment by ilopmar** [comment]
@traycho only the SSL stuff. Probably this (the code changed from when I tried): https://github.com/quarkusio/quarkus/blob/main/extensions/netty/runtime/src/main/java/io/quarkus/netty/runtime/graal/NettySubstitutions.java#L71-L354

**Comment by ilopmar** [comment]
This is what I used: [Substitutions.java.txt](https://github.com/micronaut-projects/micronaut-core/files/6591564/Substitutions.java.txt)

**Comment by ilopmar** [comment]
@traycho The GraalVM team fixed the issue and it will be included in GraalVM 21.2 (to be released on July 20th). I tried with my application and it works properly. 
Another thing fixed is that now the substitutions are not needed anymore.
